### PR TITLE
Stackless issue #290: bpo-36974, PEP 590 Vectorcall-protocol (documentation)

### PR DESCRIPTION
**Pull request by akruis**
Support the Vectorcall protocol in Stackless.

### CODE REVIEW

**Comment by akruis** [comment]
This pull request currently blocks any development, because branch "main-slp" does not compile. I'm now fairly confident that the code change is OK. But the documentation update needs to be reviewed. Therefore I'll split this pull request into two requests: a new one for the code change and this one for the documentation.

File: Doc/c-api/stackless.rst [hunk 1]
```diff
@@ -457,42 +479,67 @@ Typedef ``slp_softswitchablefunc``::
   Initialize the fields :c:member:`PyStacklessFunctionDeclarationObject.name` and
   :c:member:`PyStacklessFunctionDeclarationObject.module_name` of *sfd*.
 
-Within the body of a soft switchable extension function (or any other C-function, that obyes the stackless-protocol)
+Within the body of a soft-switchable extension function (or any other C-function, that obyes the stackless-protocol)
```
Comment (kristjanvalur) [inline Doc/c-api/stackless.rst:482]: "obeys"

File: Include/stackless_api.h [hunk 2]
```diff
@@ -400,32 +400,63 @@ PyAPI_FUNC(int) PyStackless_InitFunctionDeclaration(
...
-  care of our object, and we need no test. For example, PyObject_Call
-  and all other Py{Object,Function,CFunction}_*Call* functions use
-  STACKLESS_PROMOTE_xxx, itself, so we don't need to check further.
+  All STACKLESS_PROMOTExxx() macros are used to propagate the stackless-flag
+  from the local variable “stackless” to the global variable
+  “_PyStackless_TRY_STACKLESS”. These macros can’t be used to call a vectorcall
+  function.
+
+  The macro STACKLESS_PROMOTE_ALL() does this unconditionally. It is used for
+  cases where we know that the called function will take care of our object,
```
Comment (kristjanvalur) [inline Include/stackless_api.h:456]: "take care of our object", is confusing.  maybe fix this to say "obeys the stackless protocol by checking the `_PyStackless_TRY_STACKLESS` function and possibly returning the unwind token.  (unless I misunderstand)

File: Include/internal/pycore_slp_pystate.h [hunk 3]
```diff
@@ -70,10 +70,22 @@ typedef struct {
...
-     * flag whether the next call should try to be stackless.
-     * The protocol is: This flag may be only set if the called
+     * try_stackless: flag whether the next call should try to be stackless.
+     *
+     * Possible values:
+     *   0: don't be stackless
+     *   1: any called C-function shall try to be stackless.
+     *   other: only the C-function with address try_stackless shall try to
+     *          by stackless.
+     *
```
Comment (kristjanvalur) [inline Include/internal/pycore_slp_pystate.h:80]: This is the salient change, right? the *other* value?  Maybe it should be documented here why
 - it is necessary (something about vectorcall functions being properties)
 - how it works, in particular, how we can do it without knowing if the called function supports the protocol or not.

**Review by kristjanvalur — CHANGES_REQUESTED** [review]
The salient point seems to be the additional possible value beside 0 and 1, of the callable address.

Can you explain how the new protocol (with the function address) works?  Could that solution be used for all functions, instead of the old one having a flag in the type?

**Comment by akruis** [comment]
> The salient point seems to be the additional possible value beside 0 and 1, of the callable address.

Indeed.

> Can you explain how the new protocol (with the function address) works?

See the new comment in pycore_slp_pystate.h. It is an implementation detail, which is why I didn't describe it in the API documentation.

> Could that solution be used for all functions, instead of the old one having a flag in the type?

In theory, but it would probably hurt efficiency.  function, that only wraps
another stackless-aware function, does not need to use STACKLESS_GETARG(),
STACKLESS_PROMOTE_xxx and STACKLESS_ASSERT(). Also some stackless-aware functions
are `static inline` and have no address.

**Review by kristjanvalur — APPROVED** [review]
Great stuff.
I realize that this is API documentation.  But isn't there some general `implementation` documentation too, somewhere?  Something that doesn't require browsing .h and .c files and gives a broader overview?  I suppose that could be a good home for certain implementation details, like the mechanics of soft and hard switching.

**Comment by akruis** [comment]
> Great stuff.

Oh, thanks.

> I realize that this is API documentation. But isn't there some general `implementation` documentation too, somewhere? 
Something that doesn't require browsing .h and .c files and gives a broader overview?

That's a real issue, I'm not aware of any current 'implementation' documentation. Even worse, there is still some documentation around from times, when Stackless was really stackless. Nowadays it uses stack spilling.

Eventually Christian Tismer has some documentation.

> I suppose that could be a good home for certain implementation details, like the mechanics of soft and hard switching.

Sure.

**Pull request merged**